core/src/test/java/org/apache/calcite/plan/volcano/VolcanoRuleCallTest.java

```diff
@@ -0,0 +1,153 @@
+
+package org.apache.calcite.plan.volcano;
+
+import org.apache.calcite.plan.*;
+import org.apache.calcite.rel.logical.LogicalUnion;
+
+import org.junit.Assert;
+import org.junit.Test;
+
+import java.util.Arrays;
+
+import static org.apache.calcite.plan.volcano.PlannerTests.newCluster;
+
+public class VolcanoRuleCallTest {
+
+  public static class DummyUnionRule extends RelOptRule {
+
+    public DummyUnionRule() {
+      super(operand(LogicalUnion.class, unordered(operand(RelTypeB.class, any()))));
+    }
+
+    @Override
+    public void onMatch(RelOptRuleCall call) {
+      // do nothing
+    }
+  }
+
+  public static class RelTypeA extends PlannerTests.NoneLeafRel {
+    RelTypeA(RelOptCluster cluster, String label) {
+      super(cluster, label);
+    }
+  }
+
+  public static class RelTypeB extends PlannerTests.NoneLeafRel {
+    RelTypeB(RelOptCluster cluster, String label) {
+      super(cluster, label);
+    }
+  }
+
+
+  @Test
+  public void testUnorderedVolcanoDescending() {
+    VolcanoPlanner planner = new VolcanoPlanner();
+    planner.addRelTraitDef(ConventionTraitDef.INSTANCE);
+    planner.addRule(new DummyUnionRule());
+
+    RelOptCluster cluster = newCluster(planner);
+
+        /*
+        test plan:
+        LogicalUnion
+            RelTypeA
+            RelTypeB
+         */
+
+    RelTypeA relA1 =
+        new RelTypeA(
+            cluster,
+            "a1");
+    RelTypeB relB1 =
+        new RelTypeB(
+            cluster,
+            "b1");
+
+    int matchCounter;
+
+
+    planner.ensureRegistered(relA1, null);
+    planner.ensureRegistered(relB1, null);
+
+    LogicalUnion union = LogicalUnion.create(Arrays.asList(relA1, relB1), true);
+
+    // add the original union to planner
+    planner.ensureRegistered(union, null);
+
+    matchCounter = 0;
+    while (planner.ruleQueue.popMatch(VolcanoPlannerPhase.OPTIMIZE) != null) {
+      matchCounter++;
+    }
+    // expected: should match 1 time (incorrect)
+    Assert.assertEquals(1, matchCounter);
```

Comment:
Please add relevant message for `assertEquals`.
You do have a message in comment above, so why don't you put it into the assertion message?

If I understand your test right, the message should be something like `DummyUnionRule should fire 1 times since there's 1 LogicalUnion in the plan`. Is that right?

PS. It is not clear what `1 time (incorrect)` means. Is 1 correct or incorrect?

Reply:
1 is the correct expected result. (incorrect) is removed because it can be confusing